File: 3657-check-if-grid-can-be-cut-into-sections/check-if-grid-can-be-cut-into-sections.go
func checkValidCuts(n int, rectangles [][]int) bool {
	x := make([][]int, len(rectangles))
	y := make([][]int, len(rectangles))

	for i := range rectangles {
		x[i] = []int{rectangles[i][0], rectangles[i][2]}
		y[i] = []int{rectangles[i][1], rectangles[i][3]}
	}

	return checkNotOverlap(x) || checkNotOverlap(y)
}

func checkNotOverlap(rectangles [][]int) bool {
	sort.Slice(rectangles, func(i, j int) bool {
		return rectangles[i][0] < rectangles[j][0]
	})

	sections := 0
	maxEnd := rectangles[0][1]

	for _, rect := range rectangles {
		if maxEnd <= rect[0] {
			sections++
		}
		maxEnd = max(maxEnd, rect[1])
	}

	return sections >= 2
}

File: 3657-check-if-grid-can-be-cut-into-sections/check-if-grid-can-be-cut-into-sections.java
import java.util.Arrays;

class Solution {
    public boolean checkValidCuts(int n, int[][] rectangles) {
        for (int k = 0; k < 2; k++) {
            final int index = k;
            Arrays.sort(rectangles, (a, b) -> Integer.compare(a[index], b[index]));

            boolean first = false;
            int end = rectangles[0][k + 2];

            for (int[] rect : rectangles) {
                if (rect[k] >= end) {
                    if (first) {
                        return true;
                    }

                    first = true;
                }
                
                end = Math.max(end, rect[k + 2]);
            }
        }

        return false;
    }
}
